File: src/main/java/org/micoli/php/service/DebouncedRunnables.java
package org.micoli.php.service;

import java.util.HashMap;
import java.util.Map;
import org.jetbrains.annotations.NotNull;

public class DebouncedRunnables {
    private @NotNull final Map<String, DebouncedRunnable> debouncedActions = new HashMap<>();

    public DebouncedRunnable run(@NotNull Runnable task, @NotNull String name, long delayMillis) {
        return run(task, name, delayMillis, null);
    }

    public DebouncedRunnable run(@NotNull Runnable task, @NotNull String name, long delayMillis, Runnable callback) {
        DebouncedRunnable debouncedRunnable =
                debouncedActions.computeIfAbsent(name, k -> new DebouncedRunnable(task, delayMillis, callback));
        debouncedRunnable.run();
        return debouncedRunnable;
    }

    public void reset() {
        for (DebouncedRunnable debouncedRunnable : debouncedActions.values()) {
            debouncedRunnable.close();
        }
        debouncedActions.clear();
    }
}

File: src/main/kotlin/org/micoli/php/service/DebouncedRunnables.kt
package org.micoli.php.service

open class DebouncedRunnables {
    private val debouncedActions: MutableMap<String, DebouncedRunnable> = HashMap()

    fun run(task: Runnable, name: String, delayMillis: Long): DebouncedRunnable? {
        return run(task, name, delayMillis, null)
    }

    open fun run(
        task: Runnable,
        name: String,
        delayMillis: Long,
        callback: Runnable?
    ): DebouncedRunnable? {
        val debouncedRunnable =
            debouncedActions.computeIfAbsent(name) {
                DebouncedRunnable(task, delayMillis, callback)
            }
        debouncedRunnable.run()
        return debouncedRunnable
    }

    fun reset() {
        for (debouncedRunnable in debouncedActions.values) {
            debouncedRunnable.close()
        }
        debouncedActions.clear()
    }
}
